crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,631 @@
+        .or_else(|| {
+            let mut cursor = INTERNED_STRING_BY_ID.read().unwrap().cursor();
+            if cursor.seek_key_exact(id) {
+                Some(
+                    unsafe { cursor.val().downcast::<InternedString>() }
+                        .0
+                        .clone(),
+                )
+            } else {
+                None
+            }
+        })
+}
+
+/// Create an empty spine snapshot.
+fn empty_by_id() -> SpineSnapshot<DynOrdIndexedWSet<DynData, DynData, DynZWeight>> {
+    let factories: OrdIndexedWSetFactories<DynData, DynData, DynZWeight> =
+        BatchReaderFactories::new::<InternedStringId, SqlString, ZWeight>();
+
+    SpineSnapshot::<DynOrdIndexedWSet<DynData, DynData, DynZWeight>>::new(factories)
+}
+
+/// Build the string interner circuit.
+///
+/// Takes a stream that contains strings to be interned, and sets up a spine snapshot
+/// that maps interned string IDs back to strings. The spine snapshot is stored in the
+/// `INTERNED_STRING_BY_ID` global variable, and is updated on each step of the circuit.
+///
+// ```text
+//                                               ┌──────────┐                      ┌────────────────────────┐
+//                                               │integrate │                      │ Global interner state  │
+//                                               │  ┌───┐   │       ┌───────┐by_id │                        │
+//                                               │  │Z-1├───┼──────►│gather ├─────►│ INTERNED_STRINGS_BY_ID │
+//                                               │  └──┬┘   │       └───────┘      │ CURRENT_STEP           │
+//                                               │   ▲ │    │                      │ PINNED_STRINGS         │
+//                                               │   │ ▼    │                      └────────────────────────┘
+//   feldera_interned_strings ┌─────────┐        │  ┌┴──┐   │
+// ──────────────────────────►│map_index├────────┼─►│ + │   │
+//                            └─────────┘        │  └───┘   │
+//                                               └──────────┘
+// ```
+pub fn build_string_interner(interned_strings: Stream<RootCircuit, OrdZSet<Tup1<SqlString>>>) {
+    info!("Setting up string interner");
+
+    // Intern input strings, index them by interned string ID, and store them in a spine.
+    //
+    // The last step below `.delay_trace()` makes sure that we work with the spine snapshot
+    // from the previous step, hence any entries deleted at the current step will still be
+    // present. This is important, because even after the string is removed from the spine,
+    // it may still be used during the current step. At the same time, all newly added strings
+    // that are not in the spine are guaranteed to be in the cache.
+    let by_id = interned_strings
+        .map_index(|s| (intern_string(&s.0), s.clone()))
+        .shard()
+        .set_persistent_id(Some("feldera_interned_string_by_id"))
```

Comment:
we should probably reverse namespace `feldera_` and prevent users from creating tables/views with this prefix @mihaibudiu

Reply:
In this case, we need the user to create a view with this name. This is a temporary solution until we have compiler support for interning.